man/authd.3
.\" Copyright Georgios Portokalidis, 2004.
.\"
.\" This file may be copied under the conditions described
.\" in the GNU GENERAL PUBLIC LICENSE, Version 2, June 1991
.\" that should have been distributed together with this file.
.\"
.TH AUTHD 3 "April 22, 2004" "Version 0.8" "System Library Functions Manual"
.SH NAME
libadmctrlcl.a \- Admission control client library
.SH LIBRARY

.SH SYNOPSIS
.\"
.\" ADMISSION CONTROL CLIENT
.\"
.B #include <admctrlcl.h>
.P
struct adm_ctrl_pair
{
  char name[MAX_PAIR_NAME];
  char value[MAX_PAIR_VALUE];
.br
};
.br
typedef struct adm_ctrl_pair adm_ctrl_pair_t;
.P
struct adm_ctrl_result
{
  int PCV;
  int error;
  size_t resources_num;
.br
#ifdef WITH_RESOURCE_CONTROL
  resource_required_t required[RESOURCE_CTRL_MAX_RESOURCES];
.br
#endif
.br
};
.br
typedef struct adm_ctrl_result adm_ctrl_result_t;
.P
struct adm_ctrl_request
{
  unsigned char pubkey[MAX_PUBKEY_SIZE];
  unsigned char credentials[MAX_CREDENTIALS_SIZE];
  unsigned int nonce;
  unsigned char encrypted_nonce[MAX_ENC_NONCE_SIZE];
  size_t encrypted_nonce_len;
  unsigned int pairs_num;
  adm_ctrl_pair_t pair_assertions[MAX_PAIR_ASSERTIONS];
  unsigned int functions_num;
  unsigned char function_list[MAX_FUNCTION_LIST_SIZE];
.br
};
.br
typedef struct adm_ctrl_request adm_ctrl_request_t;
.P
typedef struct admctrlcl admctrlcl_t;
.\" ADMCTRLCL_NEW_IPC
.P
.B "admctrlcl_t *"
.br
.BI "admctrlcl_new_ipc(const char *" pathname ", int " id ","
.BI "char " persistent ", struct timeval *" timeout ","
.BI "adm_ctrl_request_t *" request ", adm_ctrl_result_t *" result ");"
.\" ADMCTRLCL_NEW_SOCKET
.P
.B "admctrlcl_t *"
.br
.BI "admctrlcl_new_socket(const char *" hostname ", int " port ","
.BI "char " persistent ", struct timeval *" timeout ","
.BI "adm_ctrl_request_t *" request ", adm_ctrl_result_t *" result ");"
.\" ADMCTRLCL_USE_SSL
.P
.B int
.br
.BI "admctrlcl_use_SSL(admctrlcl_t *" client ", const char *" key ","
.BI "const char *" ca ");"
.\" ADMCTRLCL_DESTROY
.P
.BI "void admctrlcl_destroy(admctrlcl_t *" client ");"
.\" ADMCTRLCL_SET_REQUEST
.P
.BI "void admctrlcl_set_request(admctrlcl_t *" client ", const
adm_ctrl_request_t
.BI "*" request ");"
.\" ADMCTRCL_GET_RESULT
.P
.B "adm_ctrl_result_t *"
.br
.BI "admctrlcl_get_result(admctrlcl_t *" client ");"
.\" ADMCTRCL_COMM_OPEN
.P
.BI "int admctrlcl_comm_open(admctrlcl_t *" client ");"
.\" ADMCTRLCL_COMM_CLOSE
.P
.BI "int admctrlcl_comm_close(admctrlcl_t *" client ");"
.\" ADMCTLCL_RESET
.P
.BI "void admctrlcl_reset(admctrlcl_t *" client ");"
.\" ADMCTRLCL_SUBMIT_REQUEST
.P
.BI "int admctrlcl_submit_request(admctrlcl_t *" client ");"
.\"
.\" ADMISSION CONTROL REQUEST
.\"
.P
.P
.B #include <admctrl_req.h>
.\" ADMCTRL_REQ_SET_AUTHINFO"
.P
.B void
.br
.BI "admctrl_req_set_authinfo(adm_ctrl_request_t *" request ","
.BI "const unsigned char *" pub ", const unsigned char *" creds ","
.BI "unsigned int " nonce ", const unsigned char *" enc_nonce ","
.BI "size_t " enc_nonce_len ");"
.\" ADMCTRL_REQ_ADD_NVPAIR
.P
.B int
.br
.BI "admctrl_req_add_nvpair(adm_ctrl_request_t *" request ","
.BI "const char *" name ", const char *" value ");"
.\" ADMCTRL_REQ_ADD_FUNCTION
.P
.B int
.br
.BI "admctrl_req_add_function(adm_ctr_request_t *" request ","
.BI "size_t *" fbuf_off ", const char *" fname ","
.BI "const char *" lname ", const char *" argt ", ...);"
.\" ADMCTRL_REQ_ADD_SFUNCTION
.P
.B int
.br
.BI "admctrl_req_add_sfunction(adm_ctrl_request_t *" request ","
.BI "size_t *" fbuf_off ", const char *" fname ","
.BI "const char *" lname ", const char *" argt ","
.BI "const unsigned char *" args ", size_t " args_size ");"
.\" ADMCTRL_REQ_ENCRYPT_NONCE
.P
.B size_t
.br
.BI "admctrl_req_encrypt_nonce(unsigned char **" enc_nonce ","
.BI "unsigned int " nonce ", const char *" priv ");"
.\" LINK OPTIONS
.P
.B Compile options: \-DWITH_RESOURCE_CONTROL
.br
Use this compiler flag if you are compiling with resource control.
.P
.B Link options: \-ladmctrlcl
.br
.RB "If SSL was enabled at compile time you will also need to use " -lssl "."
.SH DESCRIPTION
.\" ADMCTRCL_NEW_IPC
.B admctrlcl_new_ipc()
creates a new admission control client that will be using IPC.
.IR pathname " and " id " are used to generate the key for IPC."
.I persistent
should be non-zero if communication is to established only once, when
.B admctrlcl_comm_open()
is called and not for every submitted request.
.RB "The timeout of the " admctrlcl_submit_request() " is specified by"
.IR timeout ". 0 timeout disables it."
.RI "If " request " and " result " are not NULL, instead of allocating"
new structures for sending requests and receiving results the ones supplied are
going to be used. On success a pointer to a new client is returned, or NULL on
error. 
.IR errno " is set to ENOMEM if no memory could be allocated or is set by"
.BR ftok() " if an IPC key could not be generated."
.\" ADMCTRLCL_NEW_SOCKET
.P
.B admctrlcl_new_socket()
creates a new admission control client that will be using sockets.
.IR hostname " and " port " specify the host name and port number of the
admission control server.
.IR persistent " should be non-zero if communication is to established only
once, when
.BR admctrlcl_comm_open() " is called and not for every submitted request."
.RB "The timeout of the " admctrlcl_submit_request() " is specified by"
.IR timeout ". 0 timeout disables it."
.RI "If " request " and " result " are not NULL, instead of allocating"
new structures for sending requests and receiving results the ones supplied are
going to be used. On success a pointer to a new client is returned, or NULL on
error.
.IR errno " is set to ENOMEM if no memory could be allocated or is set by"
.BR gethostbyname() " if the server's host name could not be resolved."
.\" ADMCTRLCL_USE_SSL
.P
.B admctrlcl_use_SSL()
.RI "enables SSL for the client pointed to by " client ". It only affects
clients using sockets and haven't yet connected to a server.
.IR key " and " ca " should be the names of the files containing the client's
private key and the list of accepted certificate authorities respectively. The
later can be disabled, if set to NULL. 0 is returned on success, or -1 on
error.
.IR errno " is set to ENOPROTOOPT if SSL is not supported or the client is using
IPC, or to EPROTO if an SSL error has occurred. OpenSSL error library functions
can be used in the second case to get an error description. Check openssl(1)
and ERR_get_error(3) for more information.
.\" ADMCTRLCL_DESTROY
.P
.B admctrlcl_destroy()
.RI "destroy the admission control client pointed to by " client "."
.\" ADMCTRLCL_SET_REQUEST
.P
.B admctrlcl_set_request()
.RI "sets the request point to by " request " as the request to be sent by the
client. The request is copied to the client's request structure.
.\" ADMCTRLCL_GET_RESULT
.P
.B admctrlcl_get_result()
.RI "retrieves the result of the last " client " submission. The client's result
structure is not copied, but instead just a pointer to it is returned. If a
successful submission hasn't been made the user might receive garbage. After a
successful submission 
.I PCV
 is 1 if the request was authorised or 0 otherwise.
.IR error " is less than zero in case of an error. Possible values are listed in
.BR admctrl_errno.h "."
.IR resources_num " is set to the number of required resources by the request.
It is always set to zero when resource control is not enabled.
.IR required " is not a valid field if resource control is disabled. Otherwise
it contains the resources required by the request. It can be used to consume
the resources by using the resource control functions. Please read \'doc
RESOURCE_CONTROL.txt\' for more information.
.\" ADMCTRLCL_COMM_OPEN
.P
.B admctrlcl_comm_open()
.RI "opens/initiates communication for " client ". It does nothing for
non-persistent clients. 0 is returned on success, or -1 on failure.
.IR errno " is set either by IPC calls, or socket and SSL operations depending
on the client type. OpenSSL error library functions can be used to get an error
description in case of SSL related errors. Check openssl(1) and
ERR_get_error(3) for more information.
.\" ADMCTRLCL_COMM_CLOSE
.P
.B admctrlcl_comm_close()
.RI "closes/cleans-up communication for " client ". It does nothing for
non-persistent clients. 0 is returned on success, or -1 on failure.
.IR errno " is set either by IPC calls, or socket and SSL operations depending
on the client type. OpenSSL error library functions can be used to get an error
description in case of SSL related errors. Check openssl(1) and
ERR_get_error(3) for more information.
.\" ADMCTRLCL_RESET
.P
.B admctrlcl_reset()
.RI "resets the request and result structures for " client ". It is wise to call
this function after submitting a request and retrieving the result, and before
setting a new request for submission.
.\" ADMCTRLCL_SUBMIT_REQUEST
.P
.B admctrlcl_submit_request()
.RI "submit the request for " client ". The timeout specified when the client
was created is used. 0 is returned on success, or -1 on failure.
.IR errno " is set either by IPC calls, or socket and SSL operations depending
on the client type. OpenSSL error library functions can be used to get an error
description in case of SSL related errors. Check openssl(1) and
ERR_get_error(3) for more information.
.\" ADMCTRL_REQ_SET_AUTHINFO
.P
.B admctrl_req_set_authinfo()
sets authentication and authorisation information in request pointed to by
.IR request ". " pub " and " creds " are buffers containing the users public
key and credentials respectively.
.IR nonce " is the random integer number given to the user and " enc_nonce
.RI "should contain the " nonce " encrypted with the user's private key."
.IR enc_nonce_len " should be the length of the encrypted nonce.
.\" ADMCTRL_REQ_ADD_NVPAIR
.P
.B admctrl_req_add_nvpair()
.RI "adds a name\-value pair action in request pointed to by " request "."
.IR name " and " value " contain the name and value of the pair respectively."
0 is returned on success, or -1 if the maximum number of pairs has been reached.
.\" ADMCTRL_REQ_ADD_FUNCTION
.P
.B admctrl_req_add_function()
.RI "adds a function action in request pointed to by " request "."
.IR fbuf_off " is a pointer to the offset of the request's function
buffer used. The user does not have to manipulate this value. Initialising the
value to 0 for each new request and passing a reference to each call of
.BR admctrl_req_add_function() " is sufficient."
.IR fname " and " lname " are the function and library name respectively."
.IR argt " is the function's argument type specification string. The length of
this string is equal to the number of arguments of the function. Each character
specifies the type of the function's corresponding argument. Valid characters
are: \'s\', \'i\', \'f\', \'L\' and \'F\' as specified 
in \'admctrl_argtypes.h\'. The variable argument list should contain as many
arguments
.RI "as the length of " argt ". O is returned on success, or -1 on error."
.\" ADMCTRL_REQ_ADD_SFUNCTION
.P
.B admctrl_req_add_sfunction()
is an alternative way to add a function action to the request pointed to by
.IR request ". The arguments are similar to "
.BR admctrl_req_add_function() ", but instead of a variable length argument list
.RI "the user should provide a buffer pointed to by " args " containing a
serialised form of the function's arguments.
.IR args_size " is the size of the function's arguments serialised form. Please
read \'doc/DEVELOPERS.txt\' for more information on serialising arguments.0 is
returned on success, or -1 on error.
.\" ADMCTRL_REQ_ENCRYPT_NONCE
.P
.B admctrl_req_encrypt_nonce()
.RI "is an assisting function for end users. It encrypts the " nonce
.RI "using the private key contained in file " priv ". A buffer is allocated and
a pointer to it is returned in 
.IR enc_nonce ". The size of the buffer containing the encrypted nonce is
returned on success, or 0 on error.
.SH RETURN VALUES
All the functions allocating memory return a valid pointer on success, or NULL
on error. While all functions returning integer (except
.B admctrl_req_encrypt_nonce()
return 0 on success, or -1 on failure. A lot of functions also set
.IR errno ". Check the description section for more information."
.SH SEE ALSO
authdfe(8), authd(8), keynote(1), keynote(4), keynote(5), openssl(1),
ERR_get_error(3)
.SH AUTHOR
Georgios Portokalidis (gportoka@liacs.nl)
.SH NOTES
IMPORTANT this version of
.B libadmctrlcl.a
doesn't support marshalling, thus will not work properly when the client and the
front-end lie in machines having different byteorder or different number sizes.
Future versions of the library will support this.


ChangeLog
ADMCTRL - Admission Control Daemon - ChangeLog
--------------------------------------------

0.8.9
=====

Configuration
  * configure.ac Fixed for additional compatibility
  * src/Makefile.am Updated to reflect changes in configure.ac
  * tests/Makefile.am Updated to reflect changes in configure.ac
  * kernel/kadmctrl_req.c Removed the 'k' prefix from functions
  * kernel/kclient.c Updated to reflect changes in kernel/kadmctrl_req.c

0.8.8
=====

Configuration
  * configure.ac Added options to specify seperate keynote header and 
  library directory,

Tests
  * tests/enc_nonce Encrypts a nonce and prints it out in hex.
  * kernel/kclient.c Kernel driver test client.

Kernel Driver
  * kernel/kadmctrl_req.c Admission control request manipulation function
  for the kernel.

0.8.7
=====

Resource control
  * arith_parser.c updated postfix_expr_parse() & infix_expr_parse() to
  successfully parse an empty string and return 0 as a result.

0.8.6
=====

Documentation
  Added authd_setup(8) man page.
  Updated authd(3) man page.

Admission control client library
  Function admctrl_req_set_authinfo() now accepts buffers
  containing the public key and credentials instead of the files containing 
  them.
  Removed function admctrl_req_encrypt_nonce(). It's an end-user assisting
  function. Code provided as a C snippet in 'doc/encrypt_nonce.c'.

0.8.5
====

Documentation
  Added more documentation under doc/.
  Written tests/README.

Configuration
  Corrected configure.ac, src/Makefile.am and tests/Makefile.am to link only
  necessary libraries.

Tests
  Updated client to support resource control and test function type arguments.

0.8.4
=====

Documentation
  Added authd(3) man page for developers and updated documentation in doc/.

Configuration
  Added option --disable-authd to enable stand alone compilation of the client
  library.

0.8.3 
=====

Admission control front-end
  Added support for linux kernel module device driver.
  Renamed front-end executable to 'authdfe'.

Configuration
  Changed --without-clientlib to --disable-clientlib.
  Changed --with-frontend to --enable-front-end.
  Changed --with-keynote-prefix to --with-keynote.
  Changed --with-db-prefix to --with-berkeley-db.
  Changed --with-pthread-prefix to --with-pthread.
  Changed --with-snprintfv-prefix to --with-snprintfv.

0.8.2
=====

Added a linux kernel module device driver

0.8
===

Started using the GNU build system


src/authd.c
/* authd.c

  Copyright 2004  Georgios Portokalidis <digital_bull@users.sourceforge.net>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/ipc.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#ifdef WITH_SYSLOG
#include <syslog.h>
#else
enum { LOG_INFO, LOG_CRIT, LOG_WARNING, LOG_ERR };
#endif

#include "shm_sync.h"
#include "shm.h"
#include "adm_ctrl.h"
#include "admctrl_errno.h"
#include "admctrl_comm.h"


/*! \file authd.c
 *  \brief The authd admission control daemon
 *  \author Georgios Portokalidis
 */


/**** Global variables ****/
//! Admission control IPC communication data
static admctrl_comm_t comm;
//! Policy
static adm_ctrl_policy_t policy;
//! Executable's name, used for error reporting
static const char *exec_name;

#ifdef WITH_RESOURCE_CONTROL
#include "resource_ctrl.h"

//! Resource control db
static resource_ctrl_db_t resctrl_db;
//! Location of resource control home
static char *resource_ctrl_home = DEFAULT_RESOURCE_CTRL_HOME;
//! Database file name
static char *resource_ctrl_name = DEFAULT_RESOURCE_DBNAME;

#endif


/**** Configurable values ****/
//! Daemon process flag
static char isdaemon = 0;
//! Resource control status flag
static char resource_control = 0;
//! Verbocity flag
static char verbose = 0;
//! Filename containing the policy
static char *policy_fn = DEFAULT_POLICY_FILE;
//! Filename to access shared memory
static char *shm_fn = DEFAULT_SHM_FILE;
//! Project id to access shared memory
static char shm_pid = DEFAULT_SHM_PROJECT_ID;


/** \brief Prints messages to syslog and additionally to stdout 
	*	if the process isn't running as a daemon
	*
	*	\param type the type of the message, passed to syslog
	* \param s the actual message to print
	*/
static inline void
print_msg(int type,const char *s)
{
#ifdef WITH_SYSLOG
	syslog(type,s);
#endif
	if ( isdaemon == 0 )
		printf("%s: %s\n",exec_name,s);
}


/** \brief Shutdown the process
 *
 * Gracefully ends this process by closing the shared memory segment,
 * deallocating memory and calling exit()
 */
static void 
shutdown(int data)
{
	if ( comm.shm_id >= 0 )
		admctrl_comm_uninit(&comm);
#ifdef WITH_RESOURCE_CONTROL
	if ( resource_control )
		resource_ctrl_dbclose(&resctrl_db);
#endif
	if ( policy.assertions )
		free(policy.assertions);
	print_msg(LOG_INFO,"Exiting");
#ifdef SYSLOG
	closelog();
#endif
	exit((data == 0 )?1:0);
}


/** \brief Display usage information
	
	\param name Name of executable
*/
static void
print_usage(const char *name)
{
	printf("Usage: %s [OPTIONS]\n\n",name);
	printf("  -d, --daemon                  Run as a daemon in the background\n");
	printf("  -p, --policy  (filename)      Read policy from filename\n");
	printf("  -s, --shmpath (pathname)      Use pathname for shared memory\n");
	printf("  -i, --shmid   (id character)  Use id for shared memory\n");
#ifdef WITH_RESOURCE_CONTROL
	printf("  -D, --dbhome  (pathname)      Set resource control DB home to pathname\n");
	printf("  -b, --dbname  (name)          Set resource control DB file name\n");
	printf("  -R, --rc                      Enable resource control\n");
#endif
	printf("  -v, --verbose                 Be verbose with clients' requests\n");
	printf("  -h, --help                    Display this message\n");
}


/** \brief Parse command line arguments

	\param argc Number of arguments
	\param argv String array containing the arguments
*/
static void
parse_arguments(int argc,char **argv)
{
	int c;
	const char optstring[] = "dp:s:i:D:hvRb:";
	const struct option longopts[] = {
		{"daemon",no_argument,NULL,'d'},
		{"policy",required_argument,NULL,'p'},
		{"shmpath",required_argument,NULL,'s'},
		{"shmid",required_argument,NULL,'i'},
		{"dbhome",required_argument,NULL,'D'},
		{"dbname",required_argument,NULL,'b'},
		{"rc",no_argument,NULL,'R'},
		{"help",no_argument,NULL,'h'},
		{"verbose",no_argument,NULL,'v'},
		{"",0,NULL,0}
	};

	while( (c = getopt_long(argc,argv,optstring,longopts,NULL)) >= 0 )
		switch( c )
		{
			case 'd':
				isdaemon = 1;
				break;
			case 'p':
				policy_fn = optarg;
				break;
			case 's':
				shm_fn = optarg;
				break;
			case 'i':
				shm_pid = *optarg;
				break;
#ifdef WITH_RESOURCE_CONTROL
			case 'D':
				resource_ctrl_home = optarg;
				break;
			case 'R':
				resource_control = 1;
				break;
      case 'b':
        resource_ctrl_name = optarg;
        break;
#endif
			case 'v':
				verbose = 1;
				break;
			case 'h':
				print_usage(argv[0]);
				exit(0);
			default:
				exit(1);
		}
}


//! The main function of the process
int 
main(int argc,char **argv)
{
	int pid,i;
	adm_ctrl_request_t *auth_request;
  adm_ctrl_result_t auth_result;
#ifdef WITH_RESOURCE_CONTROL
	resource_ctrl_db_t *DB = NULL;
#endif

	parse_arguments(argc,argv);

	// Init some values
	exec_name = *argv;
	bzero(&policy,sizeof(adm_ctrl_policy_t));
	comm.shm_id = -1;


	// Generate key for shared memory communication
	if ( (comm.key = ftok(shm_fn,shm_pid)) < 0 )
	{
		fprintf(stderr,"%s: Couldn't generate shared memory key(%s,%c)\n",exec_name,shm_fn,shm_pid);
		perror("ftok");
		return 1;
	}

	// Load policy from file
	if ( adm_ctrl_load_policy(policy_fn,&policy) < 0 )
	{
		fprintf(stderr,"%s: Couldn't load policy from %s\n",argv[0],policy_fn);
		perror("adm_ctrl_load_policy");
		return 1;
	}

#ifdef WITH_RESOURCE_CONTROL
	resctrl_db.ENV = NULL;
	// Initialise resource control
	if ( resource_control )
	{
		if ( resource_ctrl_dbinit(&resctrl_db) != 0 )
		{
			fprintf(stderr,"%s: Error initialising resource control DB\n",argv[0]);
			perror("resource_ctrl_dbinit");
			resctrl_db.ENV = NULL;
			shutdown(0);
		}
		if ( resource_ctrl_dbopen(&resctrl_db,resource_ctrl_home,resource_ctrl_name,RESOURCE_DB_RDONLY,0) != 0 )
		{
			fprintf(stderr,"%s: Error opening resource control DB in %s\n",argv[0],resource_ctrl_home);
			perror("resource_ctrl_dbopen");
			resctrl_db.ENV = NULL;
			shutdown(0);
		}
		DB = &resctrl_db;
	}
#endif

	// Initialise IPC
	switch( admctrl_comm_init(&comm) )
	{
		case -ADMCTRL_COMM_SHM_ERROR:
			fprintf(stderr,"%s: Error while initialising shared memory IPC\n",argv[0]);
			perror("admctrl_comm_init");
			shutdown(0);
		case -ADMCTRL_COMM_SEM_ERROR:
			fprintf(stderr,"%s: Error initialising semaphores for IPC\n",argv[0]);
			perror("admctrl_comm_init");
			shutdown(0);
	}
	auth_request = comm.shm_addr;

	// Go daemon
	if ( getppid() != 1 && isdaemon == 1 )
	{
		if ( (pid = fork()) < 0 )
		{
			perror(argv[0]);
			shutdown(0);
		}
		else if ( pid > 0 )
			exit(0);
		if ( setpgrp() < 0 )
		{
			perror(argv[0]);
			shutdown(0);
		}
		chdir("/");
		umask(0);

#if DEBUG == 0
		for(i = 0 ; i < NOFILE ; i++)
			close(i);
#else
    printf("DEBUG messages enabled\n");
#endif
	}

	signal(SIGTERM,shutdown);
	if ( isdaemon == 0 )
	{
		signal(SIGINT,shutdown);
		signal(SIGQUIT,shutdown);
	}

#ifdef SYSLOG
	openlog(SYSLOG_PREPEND,LOG_CONS|LOG_PID,LOG_AUTHPRIV);
#endif

	print_msg(LOG_INFO,"Running");

	// Start processing data
	while( shm_data_wait(comm.sem_id) == 0 )
	{
		bzero(&auth_result,sizeof(adm_ctrl_result_t));
		if ( (i = adm_ctrl_authenticate(auth_request)) == 1 )
		{
#ifdef WITH_RESOURCE_CONTROL
			switch( adm_ctrl_authorise(auth_request,&policy,&auth_result,DB) )
#else
			switch( adm_ctrl_authorise(auth_request,&policy,&auth_result) )
#endif
			{
				case ADMCTRL_MEMORY_ERROR:
					print_msg(LOG_CRIT,"runned out of memory");
					break;
				case ADMCTRL_INTERNAL_ERROR:
					print_msg(LOG_CRIT,"unexpected internal error");
					break;
			}
		}
		else
			auth_result.error = i;

		memcpy(comm.shm_addr,&auth_result,sizeof(adm_ctrl_result_t));
		if ( shm_result_ready(comm.sem_id) < 0 )
			break;

		// Verbose error reporting
		if ( verbose )
		{
			if ( i == ADMCTRL_AUTHENTICATION_ERROR )
				print_msg(LOG_WARNING,"request authentication failed");
			else if ( i < 0 )
				print_msg(LOG_ERR,"error while authenticating request");
			else if ( auth_result.PCV < 1 )
				print_msg(LOG_ERR,"request authorisation failed");
			else
				print_msg(LOG_INFO,"request authenticated & authorised successfully");
		}
	}

	print_msg(LOG_CRIT,"IPC failed");
	
	shutdown(0);
	return 0;// Never reaches this, but it makes gcc happy ;)
}
